HW0/P5/P5.txt

```diff
@@ -1 +1,92 @@
+1. Get a bag from another cashier; or
+2. Get the total of that bag from another cashier.
+
+In either case, the time taken is 1 second so for this problem it does not matter. For simplicity, assume each layer is composed of different cashiers so the cashiers on the base of the tree (at the leaves) each go and obtain 1 bag from us. They then communicate (possibly concurrently since that is allowed) the total of their bag to the cashiers at the level above them.
+
+It takes 256 seconds for all cashiers at the base layer to obtain bags. They then communicate the bag totals to the 128 cashiers on the 7th layer in 1 second. Then, the analysis proceeds as in question 5 with calculation and communication taking 15 seconds. It is possible for cashiers to start calculating their totals as soon as they receive their bags, but we will reach a bottleneck since only after 256 seconds do the cashiers on the base layer all have a bag. Therefore, the total time taken is the same regardless of whether they start calculating straight away or only after all base layer cashiers have bags.
+
+The total time taken is 256 + 1 + 8 + 7 = 272.
+
+### 2 bags at a time
+
+On the 7th layer, there are 128 cashiers who each require two bags. This will take 128 seconds if we can only hand bags out at a rate of 2 bags per second (per cashier). Whilst cashiers could get started calculating totals of the upper layers while we are still handing out bags, there is going to be a bottle neck since only after 129 seconds can the 7th layer be entirely completed.
+
+So, including communication time, it takes 128 + 8 + 7 = 143 seconds if we hand each cashier 2 bags at a time.
+
+# Better strategies
+
+The strategy so far seems to be hindered significantly by the time it takes us to hand the bags to the cashiers. Moving bags is slow and must be performed sequentially, but it seems that communication between cashiers is easier and can occur concurrently.
+
+Suppose there is an incredible communicator who can simultaneously communicate the labels on the bags to all cashiers on the 7th layer (without moving any of the bags.) Then it would make sense for us to give this person all 256 bags in 1 second. This person then takes 2 seconds to concurrently communicate 2 numbers to each cashier on the 7th layer (128 cashiers.)
+
+The total time taken would then be 1 + 2 + (8 + 7) = 18 seconds.
+
+Even if we cannot assume such superhuman abilities, we can still do better by giving cashiers more bags and then having them communicate or give bags to other cashiers. Let’s assume we need to distribute 6 bags to 3 cashiers (A, B, C) such that each cashier has 2 bags.
+
+1 second: We give cashier A 4 bags.
+2 seconds: We give cashier B 2 bags. Cashier A gives C 2 bags.
+
+The total time is now 2 seconds instead of 3 seconds since we benefit from A’s help. The point is, we want to reduce the amount of times we need to hand out bags by increasing the number of bags we give out each time.
```

Comment:
I guess for this problem, more employees means more handing&communication time but less counting time.
So the problem is a trade-off. Not quite sure that communication between cashiers is easier and takes less time. but the idea of superhuman ability is interesting.